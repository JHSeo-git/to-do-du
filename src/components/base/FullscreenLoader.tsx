import React from 'react';
import styled from 'styled-components';
import Spinner from 'components/common/Spinner';
import logo from 'static/images/logo.svg';

const FullscreenLoaderWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(200, 200, 200, 0.3);
  z-index: ${(props) => props.theme.zIndex.modal};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  font-size: ${(props) => props.theme.fontSizes[11]};
  color: ${(props) => props.theme.primaryColor};
`;

const LogoImg = styled.img`
  height: 8rem;
`;

const FullscreenLoader = () => {
  return (
    <FullscreenLoaderWrapper>
      <LogoImg src={logo} alt="logo" />
      <Spinner />
    </FullscreenLoaderWrapper>
  );
};

export default FullscreenLoader;
