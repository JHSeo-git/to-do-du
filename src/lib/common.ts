export const pressedKey = (fn: () => void) => (e: React.KeyboardEvent) => {
  if (e.key === 'Enter') {
    fn();
  }

  return null;
};
