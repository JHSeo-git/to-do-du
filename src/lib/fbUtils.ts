import firebase from "firebase/app";
import { firebaseInstance } from "fBase";
import { cancelled, take, put } from "redux-saga/effects";
import { EventChannel, eventChannel, buffers } from "redux-saga";

export const getStringToAuthProvider = (provider: string) => {
  let authProvider: firebase.auth.AuthProvider | null = null;
  if (provider === "Github") {
    authProvider = new firebaseInstance.auth.GithubAuthProvider();
  } else if (provider === "Google") {
    authProvider = new firebaseInstance.auth.GoogleAuthProvider();
  } else if (provider === "Facebook") {
    authProvider = new firebaseInstance.auth.FacebookAuthProvider();
  }
  return authProvider;
};

export function makeChannel(
  collection: firebase.firestore.CollectionReference,
  buffer = buffers.none(),
  snapshotListenOptions: firebase.firestore.SnapshotListenOptions | undefined
) {
  const ref = collection;

  const channel = eventChannel((emit) => {
    const unsubscribe = snapshotListenOptions
      ? ref.onSnapshot(snapshotListenOptions, emit)
      : ref.onSnapshot(emit);

    // Returns unsubscribe function
    return unsubscribe;
  }, buffer);

  return channel;
}

// TODO: Refactoring
// export interface SyncChannelOptions {
//   onSuccess: typeof asyncSyncTodos.success;
//   onFailure: typeof asyncSyncTodos.failure;
//   snapshotListenOptions: firebase.firestore.SnapshotListenOptions | undefined;
//   transform: any;
// }

export function* syncChannel(channel: EventChannel<any>, options: any) {
  const { onSuccess, onFailure, transform } = options;

  try {
    while (true) {
      const data = yield take(channel);
      const transformedData = transform ? transform(data) : data;
      yield put(onSuccess(transformedData));
    }
  } catch (err) {
    if (onFailure) yield put(onFailure(err.message));
    else
      console.error(
        "The following error has been ignored because no `failureActionCreator` has been set:",
        err
      );
  } finally {
    if (yield cancelled()) {
      channel.close();
    }
  }
}
