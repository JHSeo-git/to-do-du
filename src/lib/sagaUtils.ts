import { AsyncActionCreatorBuilder, createAsyncAction } from 'typesafe-actions';
import { put, call, Effect, SimpleEffect } from 'redux-saga/effects';

type StripEffects<T> = T extends IterableIterator<infer E>
  ? E extends Effect | SimpleEffect<any, any>
    ? never
    : E
  : never;

type DecideReturn<T> = T extends Promise<infer R>
  ? R // If it's a promise, return the promised type.
  : T extends IterableIterator<any>
  ? StripEffects<T> // If it's a generator, strip any effects to get the return type.
  : T; // Otherwise, it's a normal function and the return type is unaffected.

export type CallReturnType<T extends (...args: any[]) => any> = DecideReturn<
  ReturnType<T>
>;

interface AsyncAction {
  REQUEST: string;
  SUCCESS: string;
  FAILURE: string;
}

export const asyncActionCreator = (actionName: string): AsyncAction => {
  const asyncTypeAction: string[] = ['_REQUEST', '_SUCCESS', '_FAILURE'];

  return {
    REQUEST: actionName + asyncTypeAction[0],
    SUCCESS: actionName + asyncTypeAction[1],
    FAILURE: actionName + asyncTypeAction[2],
  };
};

export const asyncAction = <T, P, J>(asyncAction: AsyncAction) => {
  return createAsyncAction(
    asyncAction.REQUEST,
    asyncAction.SUCCESS,
    asyncAction.FAILURE
  )<T, P, J>();
};

type PromiseCreatorFunction<P, T> =
  | ((payload: P) => Promise<T>)
  | (() => Promise<T>);

export default function createAsyncSaga<
  RequestType,
  RequestPayload,
  SuccessType,
  SuccessPayload,
  FailureType,
  FailurePayload
>(
  asyncAction: AsyncActionCreatorBuilder<
    [RequestType, [RequestPayload, undefined]],
    [SuccessType, [SuccessPayload, undefined]],
    [FailureType, [FailurePayload, undefined]]
  >,
  asyncFunction: PromiseCreatorFunction<RequestPayload, SuccessPayload>,
  successFunc?: any,
  failureFunc?: any
) {
  return function* saga(action: ReturnType<typeof asyncAction.request>) {
    try {
      const result: SuccessPayload = yield call(
        asyncFunction,
        (action as any).payload
      );
      yield put(asyncAction.success(result));
      if (successFunc) {
        yield call(successFunc, result);
      }
    } catch (e) {
      yield put(asyncAction.failure(e));
      if (failureFunc) {
        yield call(failureFunc, e);
      }
    }
  };
}
