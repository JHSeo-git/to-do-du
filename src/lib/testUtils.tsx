import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';
import { ThemeProvider } from 'styled-components';
import HelmetGlobal from 'components/base/HelmetGlobal';
import store from 'store';
import { GlobalStyle } from 'styles/globalStyles';
import { theme } from 'styles/theme';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Router } from 'react-router-dom';

interface Props {
  children: React.ReactNode;
}

interface RouteProps {
  route?: string;
  history?: MemoryHistory;
}

const AllTheProviders = ({ children }: Props) => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme.lightTheme}>
        <HelmetProvider>
          <HelmetGlobal />
          <GlobalStyle />
          {children}
        </HelmetProvider>
      </ThemeProvider>
    </Provider>
  );
};

const customRender = (ui: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const withRouterRender = (
  ui: any,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: RouteProps = {},
  options?: any
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>, {
      wrapper: AllTheProviders,
      ...options,
    }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
export { withRouterRender as renderWithRouter };
