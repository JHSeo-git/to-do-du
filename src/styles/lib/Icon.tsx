import styled from "styled-components";
import { lighten, darken } from "polished";
import { FaRegCircle, FaPlus, FaMinusCircle } from "react-icons/fa";
import { BiArrowFromLeft } from "react-icons/bi";

export const PlusIcon = styled(FaPlus)`
  color: ${(props) => props.theme.primaryColor};
`;

export const RegIcon = styled(FaRegCircle)`
  color: ${(props) => props.theme.grayDarkColor};
`;

export const DeleteIcon = styled(FaMinusCircle)`
  color: ${(props) => props.theme.alertColor};
  cursor: pointer;
  transition: all 0.2s linear;
  &:hover {
    color: ${(props) => lighten(0.1, props.theme.alertColor)};
  }
  &:active {
    color: ${(props) => darken(0.1, props.theme.alertColor)};
  }
`;

export const CloseIcon = styled(BiArrowFromLeft)`
  color: ${(props) => props.theme.grayDarkColor};
  cursor: pointer;
  transition: all 0.2s linear;
  &:hover {
    color: ${(props) => lighten(0.2, props.theme.grayDarkColor)};
  }
  &:active {
    color: ${(props) => darken(0.2, props.theme.grayDarkColor)};
  }
`;
