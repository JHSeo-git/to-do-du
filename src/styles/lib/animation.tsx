import { css, keyframes } from 'styled-components';

const fadeInKeyframes = keyframes`
    0%{
        opacity: 0;
    }
    100%{
        opacity: 1;
    }
`;

export const fadeInWithDelay = (delay = 0) => css`
  animation: ${fadeInKeyframes} 0.2s linear forwards;
  animation-delay: ${delay};
`;
