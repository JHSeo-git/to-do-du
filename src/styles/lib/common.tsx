import { css } from "styled-components";

export const whiteBox = css`
  border-radius: 4px;
  background: ${(props) => props.theme.whiteColor};
`;
