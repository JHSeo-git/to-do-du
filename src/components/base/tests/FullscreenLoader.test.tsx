import React from 'react';
import FullscreenLoader from 'components/base/FullscreenLoader';
import { render } from 'lib/testUtils';

describe('<FullscreenLoader />', () => {
  const renderFunc = () => {
    const utils = render(<FullscreenLoader />);
    return { ...utils };
  };

  it('matches snapshot', () => {
    const utils = renderFunc();
    expect(utils.container).toMatchSnapshot();
  });

  it('should render logo icon', () => {
    const { getByAltText } = renderFunc();
    const logoImg = getByAltText('logo');

    expect(logoImg).toBeTruthy();
  });

  it('should render loading spinner', () => {
    // TODO: Spinner render testing
  });
});
