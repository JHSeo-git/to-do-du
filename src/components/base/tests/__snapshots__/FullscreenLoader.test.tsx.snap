// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FullscreenLoader /> matches snapshot 1`] = `
<div>
  <div
    class="sc-bdfBwQ gABrpj"
  >
    <img
      alt="logo"
      class="sc-gsTCUz PNRc"
      src="logo.svg"
    />
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 50 50"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
        stroke-width="2"
      >
        <circle
          cx="25"
          cy="25"
          r="1"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="spline"
            dur="1.8s"
            keySplines="0.165, 0.84, 0.44, 1"
            keyTimes="0; 1"
            repeatCount="indefinite"
            values="1; 20"
          />
          <animate
            attributeName="stroke-opacity"
            begin="0s"
            calcMode="spline"
            dur="1.8s"
            keySplines="0.3, 0.61, 0.355, 1"
            keyTimes="0; 1"
            repeatCount="indefinite"
            values="1; 0"
          />
        </circle>
        <circle
          cx="25"
          cy="25"
          r="1"
        >
          <animate
            attributeName="r"
            begin="-0.9s"
            calcMode="spline"
            dur="1.8s"
            keySplines="0.165, 0.84, 0.44, 1"
            keyTimes="0; 1"
            repeatCount="indefinite"
            values="1; 20"
          />
          <animate
            attributeName="stroke-opacity"
            begin="-0.9s"
            calcMode="spline"
            dur="1.8s"
            keySplines="0.3, 0.61, 0.355, 1"
            keyTimes="0; 1"
            repeatCount="indefinite"
            values="1; 0"
          />
        </circle>
      </g>
    </svg>
  </div>
</div>
`;
